### PR TITLE
Add support for API v1.8 features callback and hostAndPort (run project)

### PR DESCRIPTION
**Pull request by waxtell**
Hello!

I hope this ~~note~~ PR finds you well!

First and foremost, thank you very much for providing this tool.  The tool is an integral component of my CI/CD process and I am truly grateful for your efforts.

Second, I apologize in advance for the unsolicited PR.  I didn't want to create an issue and I could not locate "guidelines for contributing".

Prior to version 1.8 of the TestEngine API you could only specify environment - not endpoint.  With version 1.8 the hostAndPort feature was added and provided a mechanism for routing test requests to a specific endpoint.  My PR includes support for hostAndPort as well as callback.

Thank you again for your efforts and your consideration.

Best Regards,
Will

### CODE REVIEW

**Comment by bsubba** [comment]
Hello @waxtell, Thank you for your time and effort. We will take a look into it soon.It may take some time 
Thank you very much.

**Comment by akkikumar72** [comment]
@waxtell - We are reviewing your PR, mind fixing conflicts?

**Comment by waxtell** [comment]
@akkikumar72 done and done!  Thanks!

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "testengine-cli",
-  "version": "1.5.0",
+  "version": "1.8.0",
```
Comment (brahmdev) [inline package.json:3]: imho, I think this should be 1.6.0 instead
Because the versioning of TestEngine API and TestEngine CLI are not in sync.

Comment (waxtell) [reply package.json:3]: Was wondering the same as well...  happy to make the change.

File: bin/run_functions.js [hunk 2]
```diff
@@ -242,7 +256,7 @@ function executeProject(filename, project, options) {
     async.series([
             // First create a zip file, if needed.
             //
-        function (callback) {
+            function (callback) {
```
Comment (akkikumar72) [inline bin/run_functions.js:259]: Can we sync indentation to the rest of the functions?

File: bin/run_functions.js [hunk 3]
```diff
@@ -406,7 +419,7 @@ function executeProject(filename, project, options) {
                         callback();
                     });
             },
-        function (callback) {
+            function (callback) {
```
Comment (akkikumar72) [inline bin/run_functions.js:422]: Same as above ^

File: bin/run_functions.js (hunk 2 above)
Comment (waxtell) [reply bin/run_functions.js:259]: Absolutely!  Done and done.

File: bin/run_functions.js (hunk 3 above)
Comment (waxtell) [reply bin/run_functions.js]: Both indentation inconsistencies should now be corrected.

**Review by akkikumar72 — APPROVED** [review]
@waxtell Good job, Thanks for your contribution.

File: README.md [hunk 4]
```diff
@@ -83,6 +83,9 @@ table includes the currently implemented options.
 | tags | "smoketest,regression"  | See description above|
 | output | c:\\temp\\reports  | Directory to store reports in.|
 | reportFileName |  N/A | File name for the report. If omitted, the report will get a name based on the test job ID|
+| callback | http://localhost:8080 | The URL, to which the results will be posted|
+| environment | dev | Specify the The target environment for test job. Can not be used with **endpoint**|
```
Comment (RahulPundir83) [inline README.md:87]: **endpoint** should be called hostAndPort, the way it is in our API

File: README.md [hunk 5]
```diff
@@ -83,6 +83,9 @@ table includes the currently implemented options.
...
+| endpoint | localhost:8080 | The endpoint to be used for HTTP requests sent by this test, in the format host:[port]. Can not be used with **environment**|
```
Comment (RahulPundir83) [inline README.md:88]: **endpoint** should be called hostAndPort, the way it is in our API

File: bin/run_functions.js [hunk 6]
```diff
@@ -29,6 +29,8 @@ module.exports.dispatcher = function (args) {
         'testsuite',
         'securitytest',
         'tags',
+        'callback',
+        'endpoint',
```
Comment (RahulPundir83) [inline bin/run_functions.js:33]: **endpoint** should be called hostAndPort, the way it is in our API

File: bin/run_functions.js [hunk 7]
```diff
@@ -57,7 +59,7 @@ module.exports.dispatcher = function (args) {
 function printModuleHelp() {
     util.error("Usage: testengine run <command>");
     util.error("Commands: ");
-    util.error("   project [testsuite=<name>] [async] [skipdeps] [priorityJob] [testcase=<name>] [securitytest=<name>] [timeout=<seconds>] [tags=(tag1,tag2)] [output=<directory>] [printReport] [reportFileName=<filename>] [format=junit/excel/json/pdf] [environment=<environment name>]");
+    util.error("   project [testsuite=<name>] [async] [skipdeps] [priorityJob] [testcase=<name>] [securitytest=<name>] [timeout=<seconds>] [tags=(tag1,tag2)] [output=<directory>] [printReport] [reportFileName=<filename>] [format=junit/excel/json/pdf] [environment=<environment name>] [endpoint=<host:port>] [callback=<url>]");
```
Comment (RahulPundir83) [inline bin/run_functions.js:62]: **endpoint** should be called hostAndPort, the way it is in our API

File: bin/run_functions.js [hunk 8]
```diff
@@ -67,11 +69,17 @@ function conflictingOptionsCheck(options) {
...
+
     if ('tags' in options) {
         if (('testsuite' in options) || ('testcase' in options)) {
-            util.printErrorAndExit('Error: tags cannot be used together with testcase/testsuite ');
+            util.printErrorAndExit('Error: tags cannot be used together with testcase/testsuite');
         }
     }
+
+    if(('endpoint' in options) && ('environment' in options)) {
+        util.printErrorAndExit('Error: environment cannot be used together with endpoint');
```
Comment (RahulPundir83) [inline bin/run_functions.js:80]: **endpoint** should be called hostAndPort, the way it is in our API

File: bin/run_functions.js [hunk 9]
```diff
@@ -177,6 +185,12 @@ function getQueryStringFromOptions(options) {
                 queryStringPart = 'tags=' + encodeURI(tags);
                 break;
             }
+            case 'callback':
+                queryStringPart = 'callback=' + encodeURI(options[key]);
+                break;
+            case 'endpoint':
```
Comment (RahulPundir83) [inline bin/run_functions.js:191]: **endpoint** should be called hostAndPort, the way it is in our API

**Review by RahulPundir83 — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (waxtell) [reply README.md]: All occurrences of endpoint have been replaced with hostAndPort.

File: README.md (hunk 5 above)
Comment (waxtell) [reply README.md]: All occurrences of endpoint have been replaced with hostAndPort.

File: bin/run_functions.js (hunk 6 above)
Comment (waxtell) [reply bin/run_functions.js]: All occurrences of endpoint have been replaced with hostAndPort.

File: bin/run_functions.js (hunk 7 above)
Comment (waxtell) [reply bin/run_functions.js]: All occurrences of endpoint have been replaced with hostAndPort.

File: bin/run_functions.js (hunk 8 above)
Comment (waxtell) [reply bin/run_functions.js]: All occurrences of endpoint have been replaced with hostAndPort.

File: bin/run_functions.js (hunk 9 above)
Comment (waxtell) [reply bin/run_functions.js]: All occurrences of endpoint have been replaced with hostAndPort.

**Review by RahulPundir83 — APPROVED** [review]

**Comment by waxtell** [comment]
Thank you very much @RahulPundir83, @bsubba, @akkikumar72 and @brahmdev!

**Pull request merged**